### PR TITLE
CAR-3109: Feature/localization status report

### PR DESCRIPTION
**Pull request by msmcconnell**
# PR Details
## Description

This PR refactors/overhauls the gnss_ndt_selector package into a new localization_manager package. This package contains new logic to implement a more complex localization state machine and status reporting mechanism. 

The PR contains the following core classes
Localizer: This is the ROS node.
LocalizationManager: This is the worker class for Localizer and contains the entry/exit/do/signaling logic for the state machine
LocalizationTransitionTable: This class contains the actual transition rules used to determine the current state based on the provided signals.
LocalizationState: Enum of current states
LocalizationSignal: Enum of possible signals which could result in state changes
LocalizerMode: Enum of possible operational modes which this node can support. 

This is the state diagram that has been implemented
LocalizationStatusStatesComplex

Initial discussion on the draft design for this state machine can be found here
https://github.com/usdot-fhwa-stol/carma-msgs/pull/72
As development progressed the need to keep support for the localizer mode selection became apparent, so the state machine acquired some additional complexity. 

I'm still working on updating the design docs. 
## Related Issue
https://github.com/usdot-fhwa-stol/carma-platform/issues/800

## Motivation and Context
Support UI display of localization status, more robust sensor failure behavior, future auto initialization behavior

## How Has This Been Tested?
Unit tests and basic local integration testing

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [x] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: localization_manager/src/LocalizationTransitionTable.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+  {
+    // How to handle the combined conditions?
+    case LocalizationSignal::GOOD_NDT_FREQ_AND_FITNESS_SCORE:
+      setAndLogState(LocalizationState::OPERATIONAL, signal);
+      break;
+    case LocalizationSignal::POOR_NDT_FREQ_OR_FITNESS_SCORE:
+      setAndLogState(LocalizationState::DEGRADED, signal);
+      break;
+    case LocalizationSignal::LIDAR_SENSOR_FAILURE:
+      setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
```
Comment (MishkaMN) [inline localization_manager/src/LocalizationTransitionTable.cpp:85]: On the state diagram, it seems like we need to check mode here?

File: localization_manager/src/LocalizationTransitionTable.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+      logDebugSignal(signal);
+      break;
+  }
+}
+void LocalizationTransitionTable::signalWhenAWAIT_MANUAL_INITIALIZATION(LocalizationSignal signal)
+{
+  switch (signal)
+  {
+    case LocalizationSignal::INITIAL_POSE:
+      setAndLogState(LocalizationState::INITIALIZING, signal);
```
Comment (MishkaMN) [inline localization_manager/src/LocalizationTransitionTable.cpp:191]: On the state diagram, it seems like we need to check the mode here?

File: localization_manager/src/LocalizationTransitionTable.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+      }
+      break;
+    case LocalizationSignal::UNUSABLE_NDT_FREQ_OR_FITNESS_SCORE:
+      if (mode_ == LocalizerMode::AUTO)
+      {
+        setAndLogState(LocalizationState::DEGRADED_NO_LIDAR_FIX, signal);
+      }
+      else
+      {
+        setAndLogState(LocalizationState::AWAIT_MANUAL_INITIALIZATION, signal);
```
Comment (MishkaMN) [inline localization_manager/src/LocalizationTransitionTable.cpp:122]: LocalizationState::AWAIT_MANUAL_INITIALIZATION,
LocalizationSignal::LIDAR_SENSOR_FAILURE
seem to have same operation, should we combine them to make fewer lines for testing, or is this way more clear?

File: localization_manager/src/LocalizationTransitionTable.cpp (hunk 2 above)
Comment (msmcconnell) [reply localization_manager/src/LocalizationTransitionTable.cpp:191]: It's not actually possible to enter AWAIT_MANUAL_INITIALIZATION while in GNSS mode, so the check is implicit.

File: localization_manager/src/LocalizationTransitionTable.cpp (hunk 3 above)
Comment (msmcconnell) [reply localization_manager/src/LocalizationTransitionTable.cpp:122]: I'll make it a fall through case

File: localization_manager/src/LocalizationTransitionTable.cpp (hunk 1 above)
Comment (msmcconnell) [reply localization_manager/src/LocalizationTransitionTable.cpp:85]: Good catch. A check is needed, but the diagram showed the wrong check. I have updated the diagram and code.

**Comment by msmcconnell** [comment]
Updated state diagram 
LocalizationStatusStatesComplex

**Pull request merged**